File: scheduler/src/cairn_scheduler/fleet.py
"""Fleet state — the Durable-Objects logic (spec §7), in plain Python so it is testable
and portable to a CF Durable Object later.

Holds: the **registry** (who's in the fleet, which block, version/quant), the **topology**
(pipeline order), **health** (heartbeats → the source the recovery loop reads), and the
**durable stream token-history** (authoritative token-IDs per in-flight stream, mirrored
from the entry node so even a dead entry node is recoverable — invariant #2). It also
selects the `RecoveryPolicy` for the drain/retry/reassign loop.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, List, Tuple

from .domain import NodeState, can_transition


class FleetError(ValueError):
    pass


@dataclass
class NodeInfo:
    node_id: str
    stage: int
    layer_start: int
    layer_end: int
    version: str          # weight-version + quant tag — drives skew detection (spec §5.1)
    state: NodeState = NodeState.PROVISIONING
    last_heartbeat: float = 0.0


class FleetState:
    def __init__(self) -> None:
        self.nodes: Dict[str, NodeInfo] = {}
        self.topology: List[str] = []                # node_ids in pipeline order
        self.token_history: Dict[str, List[int]] = {}

    # --- registry ---
    def register_node(self, node_id, stage, layer_start, layer_end, version, t=0.0) -> NodeInfo:
        if node_id in self.nodes:
            raise FleetError(f"node {node_id} already registered")
        info = NodeInfo(node_id, stage, layer_start, layer_end, version, NodeState.PROVISIONING, t)
        self.nodes[node_id] = info
        return info

    def set_state(self, node_id: str, state: NodeState) -> None:
        info = self._node(node_id)
        if not can_transition(info.state, state):
            raise FleetError(f"illegal transition {info.state.value} → {state.value} for {node_id}")
        info.state = state

    def _node(self, node_id: str) -> NodeInfo:
        if node_id not in self.nodes:
            raise FleetError(f"unknown node {node_id}")
        return self.nodes[node_id]

    # --- topology ---
    def set_topology(self, ordered_node_ids: List[str]) -> None:
        for nid in ordered_node_ids:
            self._node(nid)
        self.topology = list(ordered_node_ids)

    def pipeline(self) -> List[NodeInfo]:
        return [self.nodes[nid] for nid in self.topology]

    def is_complete_pipeline(self) -> bool:
        """Topology covers a contiguous layer range with every node ACTIVE."""
        nodes = self.pipeline()
        if not nodes:
            return False
        if any(n.state != NodeState.ACTIVE for n in nodes):
            return False
        expect = nodes[0].layer_start
        for n in nodes:
            if n.layer_start != expect:
                return False
            expect = n.layer_end + 1
        return True

    # --- health ---
    def heartbeat(self, node_id: str, t: float) -> None:
        self._node(node_id).last_heartbeat = t

    def stale_nodes(self, now: float, timeout: float) -> List[str]:
        """Nodes whose heartbeat has aged past `timeout` — the hard-crash signal that the
        supervised edges raise (spec §6) and the recovery loop reads."""
        return [
            nid for nid, n in self.nodes.items()
            if n.state == NodeState.ACTIVE and (now - n.last_heartbeat) > timeout
        ]

    # --- durable stream token-history (invariant #2) ---
    def record_tokens(self, stream_id: str, token_ids: List[int]) -> None:
        self.token_history.setdefault(stream_id, []).extend(token_ids)

    def history_of(self, stream_id: str) -> List[int]:
        return list(self.token_history.get(stream_id, []))

    def drop_stream(self, stream_id: str) -> None:
        self.token_history.pop(stream_id, None)  # freed on completion (spec §8 retention)

    # --- the drain/retry/reassign loop's policy selection (spec §5.1) ---
    def select_recovery(self, node_id: str, warm_spares: int, spare_version: str) -> Tuple[str, str]:
        node = self._node(node_id)
        if warm_spares <= 0:
            return "rebuild", "no warm spare available"
        if node.version != spare_version:
            return "rebuild", f"version skew ({node.version} != {spare_version}); activations would be garbage"
        return "reassign", "warm spare available, versions match"

    def note_eviction_warning(self, node_id: str) -> None:
        self.set_state(node_id, NodeState.DRAINING)

File: control/src/fleet.ts
// Fleet state logic (spec §7) — registry / topology / health / durable token-history
// + recovery-policy selection. Pure, in-memory, unit-testable; the FleetDO (fleet-do.ts)
// wraps this and persists to Durable Object storage. Mirrors scheduler/.../fleet.py.

export enum NodeState {
  Provisioning = "provisioning",
  Staging = "staging",
  Warm = "warm",
  Loading = "loading",
  Active = "active",
  Draining = "draining",
  Dead = "dead",
}

const TRANSITIONS: Record<NodeState, NodeState[]> = {
  [NodeState.Provisioning]: [NodeState.Staging, NodeState.Dead],
  [NodeState.Staging]: [NodeState.Warm, NodeState.Dead],
  [NodeState.Warm]: [NodeState.Loading, NodeState.Draining, NodeState.Dead],
  [NodeState.Loading]: [NodeState.Active, NodeState.Dead],
  [NodeState.Active]: [NodeState.Draining, NodeState.Dead],
  [NodeState.Draining]: [NodeState.Dead],
  [NodeState.Dead]: [],
};

export function canTransition(from: NodeState, to: NodeState): boolean {
  return TRANSITIONS[from].includes(to);
}

export interface NodeInfo {
  nodeId: string;
  stage: number;
  layerStart: number;
  layerEnd: number;
  version: string;
  state: NodeState;
  lastHeartbeat: number;
}

export class FleetError extends Error {}

export const MAX_HISTORY_PER_STREAM = 131072; // per-stream token cap (= model max_context) (M6)

export class Fleet {
  nodes = new Map<string, NodeInfo>();
  topology: string[] = [];
  tokenHistory = new Map<string, number[]>();
  private rateWindows = new Map<string, { windowStart: number; count: number }>();

  registerNode(nodeId: string, stage: number, layerStart: number, layerEnd: number, version: string, t = 0): NodeInfo {
    if (this.nodes.has(nodeId)) throw new FleetError(`node ${nodeId} already registered`);
    // L5: validate the layer range and reject overlaps (a 0-layer lm_head tail has
    // layerEnd === layerStart - 1 and occupies no layers).
    if (layerEnd < layerStart - 1) {
      throw new FleetError(`node ${nodeId}: invalid layer range [${layerStart}, ${layerEnd}]`);
    }
    if (layerEnd >= layerStart) {
      for (const n of this.nodes.values()) {
        if (n.layerEnd >= n.layerStart && layerStart <= n.layerEnd && layerEnd >= n.layerStart) {
          throw new FleetError(
            `node ${nodeId} layers [${layerStart},${layerEnd}] overlap ${n.nodeId} [${n.layerStart},${n.layerEnd}]`,
          );
        }
      }
    }
    const info: NodeInfo = {
      nodeId, stage, layerStart, layerEnd, version, state: NodeState.Provisioning, lastHeartbeat: t,
    };
    this.nodes.set(nodeId, info);
    return info;
  }

  private node(nodeId: string): NodeInfo {
    const n = this.nodes.get(nodeId);
    if (!n) throw new FleetError(`unknown node ${nodeId}`);
    return n;
  }

  setState(nodeId: string, state: NodeState, t = 0): void {
    const info = this.node(nodeId);
    if (!canTransition(info.state, state)) {
      throw new FleetError(`illegal transition ${info.state} → ${state} for ${nodeId}`);
    }
    info.state = state;
    // L4: seed the heartbeat clock when a node first goes ACTIVE, so it isn't read as
    // instantly stale (lastHeartbeat defaulted to 0 at registration).
    if (state === NodeState.Active && info.lastHeartbeat === 0) info.lastHeartbeat = t;
  }

  setTopology(orderedNodeIds: string[]): void {
    for (const id of orderedNodeIds) this.node(id);
    this.topology = [...orderedNodeIds];
  }

  pipeline(): NodeInfo[] {
    return this.topology.map((id) => this.node(id));
  }

  isCompletePipeline(): boolean {
    const nodes = this.pipeline();
    if (nodes.length === 0) return false;
    if (nodes.some((n) => n.state !== NodeState.Active)) return false;
    let expect = nodes[0].layerStart;
    for (const n of nodes) {
      if (n.layerStart !== expect) return false;
      expect = n.layerEnd + 1;
    }
    return true;
  }

  heartbeat(nodeId: string, t: number): void {
    this.node(nodeId).lastHeartbeat = t;
  }

  staleNodes(now: number, timeout: number): string[] {
    const out: string[] = [];
    for (const [id, n] of this.nodes) {
      if (n.state === NodeState.Active && now - n.lastHeartbeat > timeout) out.push(id);
    }
    return out;
  }

  recordTokens(streamId: string, tokenIds: number[]): void {
    const cur = this.tokenHistory.get(streamId) ?? [];
    // M6: bound per-stream growth. Reject the overflow rather than truncating the head —
    // replay needs the full history from token 0 (a legit stream never exceeds max_context).
    if (cur.length + tokenIds.length > MAX_HISTORY_PER_STREAM) {
      throw new FleetError(`stream ${streamId} exceeds max history (${MAX_HISTORY_PER_STREAM})`);
    }
    cur.push(...tokenIds);
    this.tokenHistory.set(streamId, cur);
  }

  // Per-key fixed-window rate limit (M7) — the gateway calls this before forwarding.
  checkRate(key: string, limit: number, windowMs: number, now: number): { allowed: boolean; count: number } {
    const w = this.rateWindows.get(key);
    if (!w || now - w.windowStart >= windowMs) {
      this.rateWindows.set(key, { windowStart: now, count: 1 });
      return { allowed: true, count: 1 };
    }
    w.count += 1;
    return { allowed: w.count <= limit, count: w.count };
  }

  historyOf(streamId: string): number[] {
    return [...(this.tokenHistory.get(streamId) ?? [])];
  }

  dropStream(streamId: string): void {
    this.tokenHistory.delete(streamId); // freed on completion (spec §8 retention)
  }

  // The drain/retry/reassign loop's policy selection (spec §5.1).
  selectRecovery(nodeId: string, warmSpares: number, spareVersion: string): { policy: "reassign" | "rebuild"; reason: string } {
    const node = this.node(nodeId);
    if (warmSpares <= 0) return { policy: "rebuild", reason: "no warm spare available" };
    if (node.version !== spareVersion) {
      return { policy: "rebuild", reason: `version skew (${node.version} != ${spareVersion})` };
    }
    return { policy: "reassign", reason: "warm spare available, versions match" };
  }
}
